Measureable channels and mixtures

95-martin-orion opened the pull request:
_Partially_ addresses #3241; qudit and non-square operator support is not part of this PR. Design is presented in [this RFC](https://tinyurl.com/cirq-custom-channel).

`KrausChannel` and `MatrixMixture` both serve two purposes:
- Provide a base type for users to create their own noisy channels (and mixtures) without creating a new type
- Allow channels (and mixtures) to capture the selected operator index in a measurement result

The changes to `act_on_state_vector_args.py` enable the second item; everything else in this PR was already possible in Cirq, but previously required users to define their own classes to make use of it.

95-martin-orion commented:
Pinging @viathor for review. This is somewhat related to the recent `cirq.channel` -> `cirq.kraus` renaming; I could see `cirq.channel` being repurposed to return a `KrausChannel` if we wanted a protocol that returns a channel-like Cirq operation.

95-martin-orion commented:
Discussed in Cirq internal sync: needs documentation to make this feature visible to users.

95-martin-orion commented:
Used `git push --force` to correct a bad merge state.

95-martin-orion commented:
It's unclear to me what's causing the `KeyError: 'language_info'` in the notebook tests. I was able to find [this issue](https://github.com/nteract/papermill/issues/519) with a similar error, but it seems to be tied to ipykernel v5.4.0 (we use v5.3.4).

95-martin-orion commented:
Re-requesting review - the failing check has been resolved now (seems to have been a notebook parsing issue)

MichaelBroughton commented on cirq-core/cirq/ops/kraus_channel.py:
```diff
@@ -0,0 +1,95 @@
+                f'Input Kraus ops of shape {kraus_ops[0].shape} does not '
+                'represent an operator over qubits.'
+            )
+        self._num_qubits = int(num_qubits)
```
Wouldn't it be impossible to get to this point if `num_qubits` wasn't already an integer ? This code seems a little strange/brittle

95-martin-orion replied:
At this point `num_qubits` is an `np.integer`; we want to store a Python `int`.

MichaelBroughton commented on cirq-core/cirq/ops/kraus_channel.py:
```diff
@@ -0,0 +1,95 @@
+                'represent an operator over qubits.'
+            )
+        self._num_qubits = int(num_qubits)
+        for i, op in enumerate(kraus_ops):
+            if not op.size == kraus_ops[0].size:
+                raise ValueError(
+                    'Inconsistent Kraus operator sizes: '
+                    f'op[0]: {kraus_ops[0].size}, op[{i}]: {op.size}'
+                )
```
Do we care if these ops are CPTP ?

![image](https://user-images.githubusercontent.com/2791105/127388967-bea9b427-febc-4ad9-9d61-3638fdca3e45.png)

where `op` is the list of kraus_ops.

95-martin-orion replied:
Added option to validate. Defaults to false due to the cost of this check.

viathor replied:
Note that the cost of this check is the same as the cost of evaluating the channel on a single input (see below for some interesting details). If this cost is too high, then perhaps we need to speed things up elsewhere. If it isn't then I'd say adding the `validate` argument may be premature optimization and we should wait until someone complains about instantiation cost.

Here is the current timing:

```
In [5]: %timeit cirq.is_cptp(kraus_ops=[np.eye(2)/2, cirq.unitary(cirq.X)/2, cirq.unitary(cirq.Y)/2, cirq.unitary(cirq.Z)/2])
212 µs ± 3.26 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

so for a single-qubit channel with maximum number of Kraus operators we can do ~5,000 checks per second. Since this is python, I'd say it's probably acceptable.

----
For every channel E, there is an adjoint channel E^\dagger. If K_1,..., K_n are Kraus operators of E, then K_1^\dagger, ..., K_n^\dagger are Kraus operators of E^\dagger. Also, one can prove that E preserves the trace if and only if E^\dagger is unital, i.e. if E^\dagger maps identity matrix to identity matrix. Substituting into the formula for E^\dagger in terms of its Kraus operators yields the identity that Michael posted above. Adjoint channels have a [nice interpretation](https://quantumcomputing.stackexchange.com/a/12927/10480).

95-martin-orion replied:
There is an open issue (#408) for defining `_kraus_` (previously `_channel_`) on `cirq.Circuit` objects. Such an operation would need to construct a combinatorially-exploding number of Kraus operators, potentially over many qubits, and would not require validation if the input channels were known to be CPTP maps. I see this as a strong motivation to keep the `validate` argument.

> If this cost is too high, then perhaps we need to speed things up elsewhere.

Do you have recommendations for this? Should these optimizations be part of this PR?

viathor replied:
Thanks for the pointer to #408. I needed this in the past and wrote the code that solve the problem. I'll clean it up, add tests and will send it out (probably next week). The solution avoids combinatorial explosion of Kraus operators entirely by using a different representation (see #4179), so I'm not concerned about validation cost in this context.

I don't have a recommendation for speeding up `cirq.is_cptp`. It looks pretty good already.

After second thought, I think there is really no harm in the extra control over whether validation happens, so feel free to keep it. That said, I do think we should make the default `validate=True`.

MichaelBroughton commented on cirq-core/cirq/ops/kraus_channel.py:
```diff
@@ -0,0 +1,95 @@
+                'represent an operator over qubits.'
+            )
+        self._num_qubits = int(num_qubits)
+        for i, op in enumerate(kraus_ops):
```
Is `i` really needed here ?

MichaelBroughton commented on cirq-core/cirq/ops/kraus_channel.py:
```diff
@@ -0,0 +1,95 @@
+                    f'op[0]: {kraus_ops[0].size}, op[{i}]: {op.size}'
+                )
+        self._kraus_ops = kraus_ops
+        if key is None:
+            self._key = None
+        elif isinstance(key, value.MeasurementKey):
+            self._key = key
+        else:
+            self._key = value.MeasurementKey(key)
```
This seems a little more intuitive
```python3
if isinstance(key, str):
  key = value.MeasurementKey(key)
self._key = key
```

95-martin-orion replied:
Done.

MichaelBroughton commented on cirq-core/cirq/ops/matrix_mixture.py:
```diff
@@ -0,0 +1,106 @@
+                f'Input mixture of shape {m0.shape} does not represent an operator over qubits.'
+            )
+        self._num_qubits = int(num_qubits)
+        for i, op in enumerate(p[1] for p in mixture):
+            if not op.size == m0.size:
+                raise ValueError(
+                    f'Inconsistent unitary sizes: op[0]: {m0.size}, op[{i}]: {op.size}'
+                )
```
Do we want to verify the matrix is unitary (similar to above) ?

95-martin-orion replied:
Done.

MichaelBroughton commented on cirq-core/cirq/ops/matrix_mixture.py:
```diff
@@ -0,0 +1,106 @@
+                    f'Inconsistent unitary sizes: op[0]: {m0.size}, op[{i}]: {op.size}'
+                )
+        self._mixture = mixture
+        if key is None:
+            self._key = None
+        elif isinstance(key, value.MeasurementKey):
+            self._key = key
+        else:
+            self._key = value.MeasurementKey(key)
```
Can tighten up (like  above).

95-martin-orion replied:
Done.

MichaelBroughton commented on cirq-core/cirq/ops/matrix_mixture.py:
```diff
@@ -0,0 +1,106 @@
+            raise ValueError(
+                f'Input mixture of shape {m0.shape} does not represent an operator over qubits.'
+            )
+        self._num_qubits = int(num_qubits)
```
Also a little confusing (above).

95-martin-orion replied:
See above (`np.integer` vs `int`)

MichaelBroughton reviewed (changes requested):
These look good up to some minor tweaks in the code. Looking through #2771 it feels like we should definitely have a flag when these are built like: `validate=True` which indicates whether or not any validation steps should be done (`linalg.is_unitary` for matrixmixture and `linalg.is_cptp` for krauschannel). Given how short a fix for #2771 is, would you mind adding something like `cirq.linalg.is_cptp` ?

viathor commented on cirq-core/cirq/protocols/json_test_data/MatrixMixture.repr:
```diff
@@ -0,0 +1,10 @@
+    ], key='flip'),
+    cirq.MatrixMixture(mixture=[
+        (0.5, np.array([[1, 0], [0, 1]], dtype=np.int64)),
+        (0.5, np.array([[0, 1], [1, 0]], dtype=np.int64))
```
nit: maybe use different matrices in the second instance of MatrixMixture?

95-martin-orion replied:
Done.

viathor commented on cirq-core/cirq/ops/matrix_mixture.py:
```diff
@@ -0,0 +1,114 @@
+from cirq.ops import raw_types
+
+
+class MatrixMixture(raw_types.Gate):
```
The class represents a channel whose Kraus operators are scalar multiples of unitary matrices. These objects are called "mixed-unitary" or "random unitary" channels in the literature (see e.g. [here](https://quantumcomputing.stackexchange.com/questions/18190/determining-whether-there-exists-an-equivalent-set-of-unitary-kraus-operators)). It would make the class easier to find and understand if it was renamed to `MixedUnitaryChannel`.

95-martin-orion replied:
Done.

viathor commented on cirq-core/cirq/ops/kraus_channel.py:
```diff
@@ -0,0 +1,102 @@
+    """A generic channel that can record the index of its selected operator.
+
+    Args:
+        kraus_ops: a list of Kraus operators, formatted as numpy array.
```
Kraus operators don't have to be square matrices. Non-square matrices appear when you have channels between Hilbert spaces of different dimension, e.g. a channel that consumes a qutrit and produces a qubit.

At minimum we should be explicit in comments and instantiation checks about whether or not non-square Kraus operators are supported. Better yet, let's make this work for non-square Kraus operators!

viathor replied:
Resolved. See answer to a different comment.

95-martin-orion replied:
Clarified this in the docstring, and added a TODO for adding qudit / non-square operator support.

viathor commented on cirq-core/cirq/ops/kraus_channel.py:
```diff
@@ -0,0 +1,102 @@
+        kraus_ops = list(kraus_ops)
+        if not kraus_ops:
+            raise ValueError('KrausChannel must have at least one operation.')
+        num_qubits = np.log2(kraus_ops[0].size) / 2
```
Many interesting channels aren't n-qubit channels. Two interesting examples that come to mind: Werner-Holevo channel for odd dimension and channels that model leakage. It would be nice to support these.

95-martin-orion replied:
AFAIK Cirq doesn't have support for qubits which change dimension over the course of a circuit. While I agree supporting these could be worthwhile, I think it's out of scope for this PR. If this is something you'd like to see, could you open an issue to track it?

viathor replied:
(I think you meant to write this in response to the other comment?)

I agree with you: it's not even clear right now whether channels with different input and output spaces should be represented as acting on one qudit and changing its dimension or discarding the input qudit and targeting a new qudit. Unfortunately, this may be hard to shoehorn into cirq's model. Definitely out of scope here.

That said, let's make it clear in the comments that this is out assumption and that as a consequence all Kraus operators are square matrices.

---

The comment about supporting non-qubit channels remains, so not marking this as resolved (but will mark the other one as resolved).

95-martin-orion replied:
Sorry, I missed that these were distinct issues. The latest change includes a note about both (i.e. KrausChannel only supports qubit-based square operators).

95-martin-orion commented:
Addressed review comments and updated the description: #3241 is not fully fixed by this change, as it does not include support for qudits or non-square operators. Qudit support should be fairly straightforward, but non-square operators will require deeper changes to Cirq.

viathor commented on cirq-core/cirq/ops/kraus_channel.py:
```diff
@@ -0,0 +1,105 @@
+            raise ValueError('KrausChannel must have at least one operation.')
+        num_qubits = np.log2(kraus_ops[0].size) / 2
+        if not num_qubits.is_integer():
+            raise ValueError(
```
Should probably also check that `kraus_ops[0].shape[0] == kraus_ops[0].shape[1]`.

95-martin-orion replied:
Done. Also updated a few adjacent checks to use `shape` instead of `size`.

viathor commented on cirq-core/cirq/ops/kraus_channel.py:
```diff
@@ -0,0 +1,105 @@
+            return NotImplemented
+        if self._key != other._key:
+            return False
+        return np.allclose(self._kraus_ops, other._kraus_ops)
```
Kraus representation is not unique, so when measurement key is unspecified this check is insufficient. Instead, we should compare, e.g. the Choi representations which determine the channel uniquely. OTOH, if the measurement key is specified, then different sets of Kraus operators that give rise to the same Choi will generally describe different channels and then the current logic almost applies. I say "almost" since there is the caveat that one could apply any complex phase factor to the Kraus operators and get equivalent `KrausChannel`.

I think we should do something like this:

```python
if self._key is None and other._key is None:
    return np.allclose(cirq.kraus_to_choi(self._kraus_ops), cirq.kraus_to_choi(other._kraus_ops))
return all(cirq.equal_up_to_global_phase(k1, k2) for k1, k2 in zip(self._kraus_ops, other._kraus_ops))
```

This assumes that if measurement key is specified then `KrausChannel` instances that differ in the order of otherwise equivalent Kraus operators are to be considered unequal. This makes sense since they generate different measurement outcomes.

It'd be good to add unit tests for both cases above. You can use for example these two sets of Kraus operators for the phase damping channel:

1. I/2 and Z/2.
2. [[1, 0], [0, 0]] and [[0, 0], [0, 1]].

viathor replied:
Yes, defining a separate method is definitely fine (and TODO is also fine), but please add a comment to tell users that `KrausChannel` instances that describe the same linear map may be unequal if they're specified via different Kraus operators.

If we go this route then we should probably leave this as is and not use `equal_up_to_global_phase`. Otherwise it feels like a half-baked attempt at "behavioral" equality (global phase freedom is just a special case of unitary freedom in Kraus representation which is given by diagonal unitaries).

95-martin-orion replied:
Added a TODO for this and left the `np.allclose` in `__eq__`.

viathor commented on cirq-core/cirq/ops/mixed_unitary_channel.py:
```diff
@@ -0,0 +1,118 @@
+    This type of object is also referred to as a mixed-unitary channel.
+
+    Args:
+        mixture: a list of (probability, unitary) pairs
```
nit: maybe also indicate that only qubit unitaries are supported.

viathor commented on cirq-core/cirq/ops/mixed_unitary_channel.py:
```diff
@@ -0,0 +1,118 @@
+        return np.allclose(
+            [m[1] for m in self._mixture],
+            [m[1] for m in other._mixture],
+        )
```
The comment about `KrausChannel.__eq__` applies here, too.

viathor commented on cirq-core/cirq/ops/mixed_unitary_channel.py:
```diff
@@ -0,0 +1,118 @@
+        return MixedUnitaryChannel(mixture=list(protocols.mixture(mixture)), key=key)
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, MixedUnitaryChannel):
```
What should be the result of equality check on a `KrausChannel` and a `MixedUnitaryChannel` that define the same underlying quantum channel?

95-martin-orion replied:
I'd be hesitant to return true for such a comparison. The two may have the same effect on the circuit (and thus these should return true for the `equivalent_behavior` method described above), but there are minor differences between the objects themselves: `KrausChannel.has_mixture()` is false even if the channel describes a mixture, and `MixedUnitaryChannel` has better performance due to not needing to calculate operator probabilities.

viathor replied:
Good point. I like the `equivalent_behavior` idea.

viathor reviewed (commented):
A few more comments, mostly about equality checks.

95-martin-orion commented:
Immediate follow-up items:
- support qudit operations with these types
- add `equivalent_channel` protocol for comparing channel Choi representations

95-martin-orion commented:
@MichaelBroughton, do you have any further comments?

MichaelBroughton reviewed (approved):
LGTM
